### PR TITLE
Modbus controller

### PR DESCRIPTION
**Pull request by martgras**
# What does this implement/fix? 

Implements a generic modbus controller to control modbus slaves. 
This PR replaces  #1256 with a generic implementation that isn't limit to EPEVER. 
#1256 required modifications to the exiting 'modbus' component while this version only needs UART.  The modbus specifics are included in the code. The existing 'modbus' component is limited  to input registers and adding support to write data to the slave required a fair amount of changes.   

The implementation has been tested with a couple of modbus slaves including EPEVER Solar Charger, Heidelberg Wallbox (implemented by @derwolff)  , PZEM-017.   The example folder contains these definitions. 

@nikito7 implemented support for the EDP BOX that seems to be popular in Portugal (see https://github.com/nikito7/edp_box_modbus/blob/dev/esphome/edpbox1.yaml  ) 

Sensors, Binary Sensors, Text Sensors and Switches are implemented. 
Writing non boolean data to the slave can be done from a lambda see the epever.yaml  for an example.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1388
  
# Test Environment

- [x] ESP32
- [x] ESP8266
- [ ] Windows
- [ ] Mac OS
- [ ] Linux

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
esphome:
  name: solarstation
  platform: ESP32
  board: esp32dev

substitutions:
  updates: 30s

wifi:
  ssid: !secret wifi_sid
  password: !secret wifi_password
  reboot_timeout: 5min

time:
  - platform: sntp
    id: sntp_time
    timezone: "CET-1CEST,M3.5.0,M10.5.0/3"
    servers: pool.ntp.org

uart:
  id: mod_bus
  tx_pin: 17
  rx_pin: 16
  baud_rate: 115200
  stop_bits: 1


modbus:
  flow_control_pin: 5
  send_wait_time: 200ms
  id: mod_bus_epever

modbus_controller:
  - id: epever
    address: 0x1
    modbus_id: mod_bus_epever
    command_throttle: 0ms
    setup_priority: -10
    update_interval: 30s

sensor:
  - platform: modbus_controller
    modbus_controller_id: epever
    id: array_rated_voltage
    name: "array_rated_voltage"
    address: 0x3000
    unit_of_measurement: "V"
    register_type: read
    value_type: U_WORD
    accuracy_decimals: 1
    skip_updates: 0
    filters:
      - multiply: 0.01

  - platform: modbus_controller
    modbus_controller_id: epever
    id: array_rated_current
    name: "array_rated_current"
    address: 0x3001
    unit_of_measurement: "A"
    register_type: read
    value_type: U_WORD
    accuracy_decimals: 2
    filters:
      - multiply: 0.01

  - platform: modbus_controller
    modbus_controller_id: epever
    id: array_rated_power
    name: "array_rated_power"
    address: 0x3002
    unit_of_measurement: "W"
    register_type: read
    value_type: U_DWORD_R
    accuracy_decimals: 1
    filters:
      - multiply: 0.01

binary_sensor:
  - platform: modbus_controller
    modbus_controller_id: epever
    id: charging_input_volt_failure
    name: "Charging Input Volt Failure"
    register_type: read
    address: 0x3201
    bitmask: 0xC000

switch:
  - platform: modbus_controller
    modbus_controller_id: epever
    id: manual_control_load
    register_type: coil
    address: 2
    name: "manual control the load"
    bitmask: 1

  - platform: modbus_controller
    modbus_controller_id: epever
    id: default_control_the_load
    register_type: coil
    address: 3
    name: "default control the load"
    bitmask: 1


text_sensor:
  - platform: modbus_controller
    modbus_controller_id: epever
    name: "rtc clock test "
    id: rtc_clock_test2
    internal: true
    register_type: holding
    address: 0x9013
    register_count: 3
    raw_encode: HEXBYTES
    response_size: 6

```

# Explain your changes

There are a couple of feature requests for general modbus support. e.g. https://github.com/esphome/feature-requests/issues/952 
https://github.com/esphome/feature-requests/issues/670

This is a new implementation allows a declarative implementation for modbus slaves using RTU. Actually support the PZEM-0xx devices that use modbus can be implemented without writing more code

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).
Working on it : https://github.com/martgras/esphome-docs .

### CODE REVIEW

**Review by nikito7 — CHANGES_REQUESTED** [review]
Run ```script/build_codeowners.py```

**Review by nikito7 — APPROVED** [review]

**Review by nikito7 — CHANGES_REQUESTED** [review]
Edit: my mistake

**Review by nikito7 — APPROVED** [review]

**Comment by nikito7** [comment]
@glmnet can you check this? Regards

**Review by nikito7 — APPROVED** [review]

File: esphome/components/modbus_controller/modbuscontroller.cpp [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+    r.first_sensorkey = first_sensorkey;
+    r.skip_updates = skip_updates;
+    r.skip_updates_counter = 0;
+    register_ranges_.push_back(r);
+  }
+  return register_ranges_.size();
+}
+
+void ModbusController::dump_config() {
+  ESP_LOGCONFIG(TAG, "EPSOLAR:");
```
Comment (syssi) [inline esphome/components/modbus_controller/modbuscontroller.cpp]:
```suggestion
  ESP_LOGCONFIG(TAG, "ModbusController:");
```

File: esphome/components/modbus_controller/modbuscontroller.h [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+class ModbusController;
+
+enum class ModbusFunctionCode {
+  CUSTOM = 0x00,
+  READ_COILS = 0x01,
+  READ_DISCRETE_INPUTS = 0x02,
+  READ_HOLDING_REGISTERS = 0x03,
+  READ_INPUT_REGISTERS = 0x04,
+  WRITE_SINGLE_COIL = 0x05,
+  WRITE_SINGLE_REGISTER = 0x6,
```
Comment (syssi) [inline esphome/components/modbus_controller/modbuscontroller.h]:
```suggestion
  WRITE_SINGLE_REGISTER = 0x06,
```

File: esphome/components/modbus_controller/modbuscontroller.h [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+  READ_EXCEPTION_STATUS = 0x7,   // not implemented
```
Comment (syssi) [inline esphome/components/modbus_controller/modbuscontroller.h]:
```suggestion
  READ_EXCEPTION_STATUS = 0x07,   // not implemented
```

File: esphome/components/modbus_controller/modbuscontroller.h [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+  DIAGNOSTICS = 0x8,             // not implemented
```
Comment (syssi) [inline esphome/components/modbus_controller/modbuscontroller.h]:
```suggestion
  DIAGNOSTICS = 0x08,             // not implemented
```

File: esphome/components/modbus_controller/modbuscontroller.cpp [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+#include <sstream>
+#include <iomanip>
+
+#include "esphome/core/log.h"
+#include "esphome/core/application.h"
+#ifdef USE_MQTT
+#include "esphome/components/mqtt/mqtt_component.h"
+#include "esphome/components/mqtt/mqtt_switch.h"
+#endif
+#include "modbuscontroller.h"
```
Comment (syssi) [inline esphome/components/modbus_controller/modbuscontroller.cpp]: Did you choose the filename by intention? I would expect `modbus_controller.{cpp,h}` here.

File: esphome/components/modbus_controller/examples/test.yaml [hunk 6]
```diff
@@ -0,0 +1,138 @@
+esphome:
```
Comment (OttoWinter) [inline esphome/components/modbus_controller/examples/test.yaml]: We should not commit these examples, they're huge files that can become outdated quickly

File: esphome/components/modbus_controller/__init__.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+CONFIG_SCHEMA = (
+    cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(ModbusController),
+            cv.Optional(CONF_CTRL_PIN): pins.output_pin,
+            cv.Optional(CONF_ADDRESS, default=0x1): cv.hex_uint8_t,
+            cv.Optional(
+                CONF_COMMAND_THROTTLE, default="0ms"
+            ): cv.positive_time_period_milliseconds,
+            cv.Optional("sensors"): cv.All(
```
Comment (OttoWinter) [inline esphome/components/modbus_controller/__init__.py]: Use constant names here like `CONF_SENSORS` - if they're not in const.py you can define them in this module.

File: esphome/components/modbus_controller/__init__.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+                    cfg[CONF_ADDRESS],
+                    cfg[CONF_OFFSET],
+                    cfg[CONF_BITMASK],
+                    cfg[CONF_VALUE_TYPE],
+                    cfg[CONF_REGISTER_COUNT],
+                    cfg[CONF_SKIP_UPDATES],
+                )
+            )
+    if config.get("binary_sensors"):
+        conf = config["binary_sensors"]
```
Comment (OttoWinter) [inline esphome/components/modbus_controller/__init__.py]: What is the difference between this and the `binary_sensor` platform you have defined?

File: esphome/components/modbus_controller/modbusbase.cpp [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+  uint8_t buffer[32];
+  uint8_t *p_ = buffer;
+
+#define LOG_BYTE(b) \
+  { \
+    if (p_ - buffer < sizeof(buffer)) \
+      *p_++ = b; \
+  }
+
+#define DUMP_LOG() \
```
Comment (OttoWinter) [inline esphome/components/modbus_controller/modbusbase.cpp]: We have the `hexencode` helper method for this

File: esphome/components/modbus_controller/modbusbase.h [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+  float get_setup_priority() const override;
+
+  void send(uint8_t address, uint8_t function_code, uint16_t start_address, uint16_t number_of_entities,
+            uint8_t payload_len = 0, const uint8_t *payload = nullptr);
+
+  void send_raw(const std::vector<uint8_t> &payload);
+
+  /** RX,TX Control pin Ref: https://github.com/greays/esphome/blob/master/esphome/components/rs485/rs485.h */
+  void set_ctrl_pin(uint8_t ctrl_pin) {
+    static GPIOPin PIN(ctrl_pin, OUTPUT);
```
Comment (OttoWinter) [inline esphome/components/modbus_controller/modbusbase.h]: No, this should be passed in as a GPIOPin directly, see for example `switch.gpio` platform

File: esphome/components/modbus_controller/readme.md [hunk 11]
```diff
@@ -0,0 +1,506 @@
+# Modbus controller component
+
+## A fork of [esphome](https://github.com/esphome/esphome) adding support for monitoring a modbus device
```
Comment (OttoWinter) [inline esphome/components/modbus_controller/readme.md]: Stale comment, and we document in the docs repo, not here.

File: esphome/components/modbus_controller/modbuscontroller.h [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+  // see calc_key how the key is contructed
+  std::map<uint64_t, SensorItem *> sensormap_;
+  // Continous range of modbus registers
+  std::vector<RegisterRange> register_ranges_;
+  // Hold the pending requests to be sent
+  std::list<std::unique_ptr<ModbusCommandItem>> command_queue_;
+  std::queue<std::unique_ptr<ModbusCommandItem>> incoming_queue_;
+  uint32_t last_command_timestamp_;
+  uint16_t command_throttle_;
+  static std::atomic_bool sending_;  // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
```
Comment (OttoWinter) [inline esphome/components/modbus_controller/modbuscontroller.h]: ESPHome is single-threaded, so atomics are unneeded (plus bools are atomic on pretty much all architectures anyway for the way it is used here)

**Review by OttoWinter — COMMENTED** [review]
Could you elaborate why this can't be put into the existing modbus integration?

I really don't think we should have two integrations implementing the same thing, as that just gets super confusing for users and developers.

The existing `modbus` is only used in a couple integrations anyway, if we need to change the base design we can do that.

File: esphome/components/modbus_controller/__init__.py (hunk 8 above)
Comment (martgras) [reply esphome/components/modbus_controller/__init__.py]: Kind of historic  :) 
The code started with hardcoded registers for an epever device and then I was convinced to turn it into a more generic component.  

There are 2 different ways to define a sensor. Either inline like here (this was the original approach I started with) 

````yaml
modbus_controller:
  - id: modbus_controller_test
    address: 0x2
    modbus_id: mod_bus1
    sensors:
      - id: battery_voltage
        name: "Battery voltage"
        address: 0x331A
        offset: 0
        modbus_functioncode: "read_input_registers"
        value_type: U_WORD
      - id: battery_current
        name: "Battery current"
        address: 0x331B
        offset: 2
        modbus_functioncode: "read_input_registers"
        value_type: U_WORD
````
or using the "traditional" way 

````yaml
switch:
  - platform: modbus_controller
    modbus_controller_id: modbus_controller_test
    id: reset_to_fabric_default
    name: "Reset to Factory Default"
    address: 0x9008
    modbus_functioncode: write_single_coil
````

Both variants generate the same C++ Code . In my opinion the first variant is easier to write if you have a lot registers.
Especially when more than one register is read in single command.  The code detects that there are 2 consecutive addresses and sends a read multiple register command instead 2 read single register commands

File: esphome/components/modbus_controller/modbuscontroller.h (hunk 12 above)
Comment (martgras) [reply esphome/components/modbus_controller/modbuscontroller.h]: Thanks for the review.   My main motivation  for not using the existing component was making it easier to use the component without having to change the existing one.   My approach here didn't work for more than one modbus device on a single bus and therefore I have kept working on an improved version.   
It's now in a new branch at https://github.com/martgras/esphome/tree/multidev

@syssi  and your suggestions are implemented in that branch . Examples and readme's removed but instead I modified the original modbus component to handle writes and not only reads.  

Implemented all suggested changes and updated PR

File: esphome/components/modbus_controller/examples/test.yaml (hunk 6 above)
Comment (martgras) [reply esphome/components/modbus_controller/examples/test.yaml]: reverted to use modbus

**Comment by hnykda** [comment]
Just curious - is there anything a random esphome user who wants this can help to get this merged? :eyes:

**Comment by clomads** [comment]
Seconding as a random ESPHome user who's looking forward to this for use with an EPEVER Tracer. Anything I can do to help?

File: esphome/components/modbus_controller/__init__.py (hunk 8 above)
Comment (jesserockz) [reply esphome/components/modbus_controller/__init__.py]: I don't think there should be multiple places to define `sensors` etc. I would prefer for the config to be the traditional ESPHome way (second option above only) with the code in subfolders and less `AUTO_LOAD`ing in the main file which will allow for cutting down on binary size, as right now just including the modbus controller will include all of the sensor, binary, text, switch code in various places.

**Comment by jesserockz** [comment]
@martgras Please take a look at https://github.com/martgras/esphome/pull/29

File: esphome/components/modbus_controller/__init__.py (hunk 8 above)
Comment (martgras) [reply esphome/components/modbus_controller/__init__.py]: changed as proposed.

File: esphome/components/modbus/modbus.cpp [hunk 13]
```diff
@@ -49,48 +55,74 @@ bool Modbus::parse_modbus_byte_(uint8_t byte) {
...
-    return (byte & 0x80) != 0x80;
-
+  /*
+   if (at == 1) {
+     if (address != waiting_for_response && waiting_for_response != 0) {
+       ESP_LOGW(TAG, "Response from wrong device got %d waiting for %d", address, waiting_for_response);
+       return false;
+     }
+   }
+   */
```
Comment (jesserockz) [inline esphome/components/modbus/modbus.cpp]: Don't leave commented code in please =)
```suggestion
```

File: esphome/components/modbus_controller/__init__.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+        _LOGGER.warning(
+            "%s has been been deprected. Component id=%00s\n        Change '%s: %s' to '%s: %s' ",
+            CONF_MODBUS_FUNCTIONCODE,
+            cfg[CONF_ID],
+            CONF_MODBUS_FUNCTIONCODE,
+            cfg[CONF_MODBUS_FUNCTIONCODE],
+            CONF_REGISTER_TYPE,
+            reg_type_name,
+        )
+    return reg_type
```
Comment (jesserockz) [inline esphome/components/modbus_controller/__init__.py]: Considering this is a new component, this cannot be deprecated already

File: esphome/components/modbus_controller/output/modbus_output.h [hunk 15]
```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include "esphome/components/output/float_output.h"
+#include "esphome/components/modbus_controller/modbus_controller.h"
+#include "esphome/components/modbus_controller/sensor/modbus_sensor.h"
```
Comment (jesserockz) [inline esphome/components/modbus_controller/output/modbus_output.h]: This wont work if the user does not have any modbuscontroller sensors as the file will not be copied to the generated project.

Individual platforms should not rely on each other, only the parent component that is in the `DEPENDENCIES`

File: esphome/components/modbus_controller/text_sensor/modbus_textsensor.cpp [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    auto val = (*this->transform_func_)(this, result, data);
+    if (val.has_value()) {
+      ESP_LOGV(TAG, "Value overwritten by lambda");
+      result = val.value();
+    } else {
+      ESP_LOGV(TAG, "publishing handled by lambda - parse and publish");
+      return;
+    }
+  }
+  this->publish_state(result);
```
Comment (jesserockz) [inline esphome/components/modbus_controller/text_sensor/modbus_textsensor.cpp]: I don't think the lambdas should be responsible for publishing the state, they should just transform the value only.

**Same for all types, not just text_sensor**

File: esphome/components/modbus_controller/output/__init__.py [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+DEPENDENCIES = ["modbus_controller"]
+CODEOWNERS = ["@martgras"]
+
+
+ModbusOutput = modbus_controller_ns.class_(
+    "ModbusOutput", cg.Component, output.FloatOutput, SensorItem
+)
+
+CONF_MODBUS_SENSOR_ID = "modbus_sensor_id"
```
Comment (jesserockz) [inline esphome/components/modbus_controller/output/__init__.py]: Is this used?

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — CHANGES_REQUESTED** [review]
While reading the docs I realised I missed this:

Number should not publish to a Sensor, the Number itself should change to the value read from the controller and should be independent.

**Comment by martgras** [comment]
If Numbers had filters that would be my preference as well. 
They are a good fit for read/write registers
But they lack filters so to decouple them from sensor I'd have  at least provide a lambda for publishing the value . Maybe another multiply property...
The best option is adding filters to numbers ( for publish and write to device)

**Comment by jesserockz** [comment]
Maybe the `lambda` options should be consistently used for converting the register value to the desired frontend value, and for things like number (and select if you decide to implement that) could have a second lambda `write_lambda` which converts to the register value.

File: esphome/components/modbus_controller/__init__.py (hunk 14 above)
Comment (martgras) [reply esphome/components/modbus_controller/__init__.py]: true but there are already a couple of users working with it from my repo. 
https://github.com/martgras/esphome/discussions/11
I will remove it once the PR is ready for merge

File: esphome/components/modbus_controller/output/modbus_output.h (hunk 15 above)
Comment (martgras) [reply esphome/components/modbus_controller/output/modbus_output.h]: fixed

File: esphome/components/modbus_controller/output/__init__.py (hunk 17 above)
Comment (martgras) [reply esphome/components/modbus_controller/output/__init__.py]: no

**Comment by nikito7** [comment]
Change link to: https://github.com/nikito7/edpbox/tree/dev/esphome

File: esphome/components/modbus_controller/text_sensor/modbus_textsensor.cpp (hunk 16 above)
Comment (martgras) [reply esphome/components/modbus_controller/text_sensor/modbus_textsensor.cpp]: I can't find anymore from which component I copied this concept :)
Basically the idea is that a return {} from the lambda means  stop processing after the lambda: "everything has been taken care of" - but  yes maybe not really required here. 
I just followed the pattern from the number lambda  where it makes sense to give you the chance to handle more complex scenarios
I'm a bit undecided: 
-  remove that code path or 
- change the log message something like  " ESP_LOGV(TAG, "Empty return from lambda - no further processing");

**Comment by nikito7** [comment]
@martgras you re-added support for 4 bytes registers?

```yaml
  - platform: modbus_controller
    name: "${dev} T1 Vazio"
    address: 0x0026
    offset: 0
    modbus_functioncode: "read_input_registers"
    value_type: U_DWORD
    accuracy_decimals: 1
    filters:
      - multiply: 0.001
    unit_of_measurement: "kWh"
    device_class: energy
    state_class: total_increasing
```

**Comment by martgras** [comment]
> you re-added support for 4 bytes registers?
> ```yaml
>  - platform: modbus_controller
>     name: "${dev} T1 Vazio"
>     address: 0x0026
>     offset: 0
>     modbus_functioncode: "read_input_registers"
>     value_type: U_DWORD
>     accuracy_decimals: 1
>     filters:
>       - multiply: 0.001
>     unit_of_measurement: "kWh"
>     device_class: energy
>     state_class: total_increasing
> ```

Yes  but then you have to set register_count: 1.
The code assumes 2 registers for DWORD therefore you have to override it
(Please have a look at https://github.com/martgras/esphome/blob/testing/esphome/components/modbus_controller/readme.md#note-sep15 since there a couple of other changes)

```yaml
 - platform: modbus_controller
    name: "${dev} T1 Vazio"
    address: 0x0026
    register_type: "read"
    value_type: U_DWORD
   register_count: 1
   accuracy_decimals: 1
   filters:
     - multiply: 0.001
   unit_of_measurement: "kWh"
   device_class: energy
   state_class: total_increasing
> ```

File: esphome/components/modbus_controller/text_sensor/modbus_textsensor.cpp (hunk 16 above)
Comment (jesserockz) [reply esphome/components/modbus_controller/text_sensor/modbus_textsensor.cpp]: Like I said, I would prefer if the lambda was only used for transformation, and publishing was done in the component code. I see no real reason to have a user publish inside the lambda and will end up causing confusion/issues.

Comment (martgras) [reply esphome/components/modbus_controller/text_sensor/modbus_textsensor.cpp]: ok will remove it

**Comment by martgras** [comment]
> Maybe the `lambda` options should be consistently used for converting the register value to the desired frontend value, and for things like number (and select if you decide to implement that) could have a second lambda `write_lambda` which converts to the register value.

Technically agreed  but unsure about the naming

For incoming data (filter) I keep the name 'lambda'
For number , output and (in the future select the name) will be 'write_lambda' for  write to device actions.

ok?

**Review by nikito7 — APPROVED** [review]
From my needs. All is working

File: esphome/components/modbus_controller/__init__.py (hunk 14 above)
Comment (martgras) [reply esphome/components/modbus_controller/__init__.py]: removed

File: esphome/components/modbus_controller/switch/modbus_switch.cpp [hunk 18]
```diff
@@ -0,0 +1,70 @@
+
+#include "modbus_switch.h"
+#include "esphome/core/log.h"
+namespace esphome {
+namespace modbus_controller {
+
+static const char *const TAG = "modbus_controller.switch";
+
+void ModbusSwitch::setup() {
+  // value isn't required
```
Comment (jesserockz) [inline esphome/components/modbus_controller/switch/modbus_switch.cpp]:
```suggestion
  // value isn't required
```

**Review by jesserockz — APPROVED** [review]
LGTM! Great job =)

**Comment by reaper7** [comment]
how this implementation will affect the SDM energy meter component ?

**Comment by jesserockz** [comment]
> how this implementation will affect the SDM energy meter component ?

It does not affect it. This is a new component that allows generic access to any modbus device without building it in native ESPHome code. Native ESPHome components will still provide a minimal/less confusing way for people to use those devices though (like SDM)

**Pull request merged**